### PR TITLE
Add etcd-cfn-signal app.

### PR DESCRIPTION
**Pull request by obmarg**
Why do we need this change?
---

The old cfn-signal app did a good job of singalling to cloudformation
when the machine had finished starting up, so it could continue with a
rolling deploy.

However, it had one potential problem: it only waited for etcd to start
up, not for it to actually be ready.  This could lead to cloudformation
_thinking_ that the machines are all ready when all the new ones are
actually failing to join the existing cluster.  If this happened, it
would eventually replace all of our _working_ machines with broken ones.

Obviously this is not good.

What effects does this change have?
---

This adds a new app called `etcd-cfn-signal` that will specifically
check the health of the local etcd instance before signalling
cloudformation.

This should mean that any new instances don't get marked as good until
the point that they've actually joined the cluster.  In the event that a
machine fails to join the cluster the signal should never be sent, and
cloudformation will rollback to the old working config.  Even if that
rollback fails, we should at least not replace _all_ the machines that
have a working config.

Implementation Details
---

Decided to write this in rust because I wanted to get a bit of
expereience with it, and see how it compares to go.  Also, the
fact that it has a fairly modern type system felt like it would make
writing correct code a bit easier.  (No null pointer exceptions, yay!)

### CODE REVIEW

File: etcd-cfn-signal/src/aws.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+const INSTANCE_METADATA_URL: &'static str = "http://169.254.169.254/latest/meta-data/";
+const STACK_ID_TAG: &'static str = "aws:cloudformation:stack-id";
+const RESOURCE_ID_TAG: &'static str = "aws:cloudformation:logical-id";
+
+
+/// Signals cloudformation that we are ready.
+///
+/// Panics on failure.
+pub fn signal_cfn() {
+    temp();
```
Comment (siame) [inline etcd-cfn-signal/src/aws.rs]: Should be better name here than temp? Can't tell what it's doing from that

File: etcd-cfn-signal/src/aws.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+/// Fetches instance metadata from the provided endpoint.
+///
+/// Endpoint should be relative to the latest/meta-data folder on the instance
+/// metadata server
+fn fetch_instance_metadata(endpoint : &str) -> Result<String, Error> {
+    let client = hyper::Client::new();
+
+    let url = String::from(INSTANCE_METADATA_URL) + endpoint;
+    let mut resp = try!(client.get(&url).send());
```
Comment (siame) [inline etcd-cfn-signal/src/aws.rs:78]: Why do you have to send after getting the URL?

File: etcd-cfn-signal/src/aws.rs (hunk 1 above)
Comment (obmarg) [reply etcd-cfn-signal/src/aws.rs]: Actually that shouldn't even be there: debugging stuff that I removed in a future commit (but that commit is in a future PR).

File: etcd-cfn-signal/src/aws.rs (hunk 2 above)
Comment (obmarg) [reply etcd-cfn-signal/src/aws.rs:78]: `client.get()` just returns a request, that you can use to add headers, data etc.

Calling send on that request does the actual sending.

**Comment by lukasz-madon** [comment]
Rust :O

File: etcd-cfn-signal/src/etcd.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+/// Polls the etcd health endpoint until it reports we're healthy.
+pub fn wait_till_healthy(url: String, options: SSLOptions) -> Result<(), Error> {
+    let client = try!(etcd_client(options));
+    loop {
+        thread::sleep(Duration::from_millis(1000));
+        println!("Checking etcd status...");
+
+        // TODO: Need to prepend /health onto the URL...
```
Comment (siame) [inline etcd-cfn-signal/src/etcd.rs]: Has this TODO been completed?

Comment (obmarg) [reply etcd-cfn-signal/src/etcd.rs]: Has now.

File: etcd-cfn-signal/src/etcd.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+/// Polls the etcd health endpoint until it reports we're healthy.
+pub fn wait_till_healthy(server_url: String, options: SSLOptions)
+                         -> Result<(), Error> {
+    let client = try!(etcd_client(options));
+    loop {
+        thread::sleep(Duration::from_millis(1000));
+        println!("Checking etcd status...");
+
+        let health_url = server_url + "/health";
+        if let Ok(response) = client.get(&url).send() {
```
Comment (siame) [inline etcd-cfn-signal/src/etcd.rs]: This is still `url` not `health_url`

Comment (obmarg) [reply etcd-cfn-signal/src/etcd.rs]: :facepalm:

Comment (obmarg) [reply etcd-cfn-signal/src/etcd.rs]: fixed again.

**Comment by siame** [comment]
LGTM 🐑

**Pull request merged**